File: java/src/main/java/com/sketches_oxide/frequency/SALSA.java
package com.sketches_oxide.frequency;

import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * SALSA - Self-Adjusting Learned Sketch Algorithm.
 *
 * <p>SALSA is a frequency estimation sketch that dynamically adjusts its counter
 * sizing based on observed data patterns. It uses self-adjusting counter widths
 * to optimize accuracy for the specific workload.
 *
 * <p><strong>Algorithm Overview:</strong>
 * <p>SALSA employs adaptive counter sizing:
 * <ul>
 *   <li>Counters start small and grow as needed</li>
 *   <li>High-frequency items get larger counters automatically</li>
 *   <li>Low-frequency items use minimal space</li>
 *   <li>Adaptation happens dynamically during updates</li>
 * </ul>
 *
 * <p><strong>Key Features:</strong>
 * <ul>
 *   <li>Self-adjusting counter sizes</li>
 *   <li>Workload-adaptive accuracy</li>
 *   <li>Better memory utilization than fixed-width counters</li>
 *   <li>Confidence metrics for estimate quality</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Frequency estimation with unknown distributions</li>
 *   <li>Adaptive heavy hitter detection</li>
 *   <li>Workload-specific optimization</li>
 *   <li>Memory-constrained environments</li>
 * </ul>
 *
 * <p><strong>Space Complexity:</strong> O(1/epsilon * ln(1/delta)) (adaptive)
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (SALSA salsa = new SALSA(0.01, 0.01)) {
 *     for (String item : stream) {
 *         salsa.update(item);
 *     }
 *
 *     System.out.println("Frequency: " + salsa.estimate("key"));
 *     System.out.println("Confidence: " + salsa.confidenceMetric());
 *     System.out.println("Adaptation level: " + salsa.adaptationLevel());
 * }
 * </pre>
 *
 * @see <a href="https://arxiv.org/abs/2004.07954">SALSA Paper</a>
 * @see CountMinSketch
 * @see CountSketch
 */
public final class SALSA extends NativeSketch implements MergeableSketch<SALSA> {

    private final double epsilon;
    private final double delta;

    /**
     * Create a new SALSA sketch.
     *
     * @param epsilon relative error bound (0 < epsilon < 1)
     *                Smaller values mean more accurate but larger space
     * @param delta failure probability (0 < delta < 1)
     *              Smaller values mean more confident but larger space
     * @throws IllegalArgumentException if parameters are invalid
     * @throws OutOfMemoryException if native memory allocation fails
     */
    public SALSA(double epsilon, double delta) {
        if (epsilon <= 0 || epsilon >= 1) {
            throw new IllegalArgumentException("epsilon must be in range (0, 1), got: " + epsilon);
        }
        if (delta <= 0 || delta >= 1) {
            throw new IllegalArgumentException("delta must be in range (0, 1), got: " + delta);
        }

        this.epsilon = epsilon;
        this.delta = delta;
        this.nativePtr = SketchOxideNative.salsa_new(epsilon, delta);

        if (this.nativePtr == 0) {
            throw new OutOfMemoryException("Failed to allocate SALSA");
        }
    }

    /**
     * Update the frequency of an item (increment by 1).
     *
     * <p>The sketch may adapt its internal counter sizing based on this update.
     *
     * @param item the item to count
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public void update(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.salsa_update(nativePtr, item);
    }

    /**
     * Update with a string item.
     *
     * @param item the string to count
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public void update(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        update(item.getBytes());
    }

    /**
     * Estimate the frequency of an item.
     *
     * <p>Returns an estimate using the adapted counter configuration.
     *
     * @param item the item to query
     * @return estimated frequency
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public long estimate(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return SketchOxideNative.salsa_estimate(nativePtr, item);
    }

    /**
     * Estimate the frequency of a string item.
     *
     * @param item the string to query
     * @return estimated frequency
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public long estimate(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return estimate(item.getBytes());
    }

    /**
     * Get the confidence metric for the current sketch state.
     *
     * <p>Higher values indicate more confident estimates. This metric
     * reflects how well the sketch has adapted to the data distribution.
     *
     * @return confidence metric (0.0 to 1.0)
     * @throws IllegalStateException if the sketch has been closed
     */
    public double confidenceMetric() {
        checkAlive();
        return SketchOxideNative.salsa_confidenceMetric(nativePtr);
    }

    /**
     * Get the current adaptation level.
     *
     * <p>Indicates how much the sketch has adapted from its initial state.
     * Higher levels indicate more adaptation has occurred.
     *
     * @return adaptation level (0 = no adaptation)
     * @throws IllegalStateException if the sketch has been closed
     */
    public int adaptationLevel() {
        checkAlive();
        return SketchOxideNative.salsa_adaptationLevel(nativePtr);
    }

    /**
     * Merge another SALSA into this one.
     *
     * <p>Both sketches must have the same epsilon and delta parameters.
     *
     * @param other the sketch to merge
     * @throws IllegalStateException if either sketch is closed
     * @throws IllegalArgumentException if parameters don't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(SALSA other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (this.epsilon != other.epsilon || this.delta != other.delta) {
            throw new IllegalArgumentException(
                    "Cannot merge SALSA sketches with different parameters: " +
                            String.format("(%.3f, %.3f) vs (%.3f, %.3f)",
                                    this.epsilon, this.delta, other.epsilon, other.delta));
        }

        SketchOxideNative.salsa_merge(this.nativePtr, other.nativePtr);
    }

    /**
     * Get the epsilon parameter.
     *
     * @return the epsilon (relative error bound)
     */
    public double epsilon() {
        return epsilon;
    }

    /**
     * Get the delta parameter.
     *
     * @return the delta (failure probability)
     */
    public double delta() {
        return delta;
    }

    /**
     * Serialize the sketch to binary format.
     *
     * @return binary representation of the sketch
     * @throws IllegalStateException if the sketch has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.salsa_serialize(nativePtr);
    }

    /**
     * Deserialize from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new SALSA instance
     * @throws IllegalArgumentException if data is invalid
     * @throws NullPointerException if data is null
     */
    public static SALSA deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.salsa_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized SALSA data");
        }

        SALSA salsa = new SALSA(0.01, 0.01);
        SketchOxideNative.salsa_free(salsa.nativePtr);
        salsa.nativePtr = ptr;
        return salsa;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.salsa_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        try {
            if (nativePtr != 0) {
                return String.format("SALSA(epsilon=%.3f, delta=%.3f, confidence=%.2f, adaptation=%d)",
                        epsilon, delta, confidenceMetric(), adaptationLevel());
            }
        } catch (IllegalStateException e) {
            // Closed
        }
        return "SALSA(closed)";
    }
}

File: dotnet/SketchOxide/src/Frequency/SALSA.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Frequency;

/// <summary>
/// SALSA (Self-Adjusting Lean Streaming Analytics) sketch for frequency estimation.
///
/// A novel sketch that dynamically adjusts counter sizes based on observed frequencies.
/// SALSA starts with small counters and expands them as needed, achieving better space
/// efficiency than fixed-width counter approaches while maintaining accuracy.
/// </summary>
/// <remarks>
/// Note: SALSA does not support merging as the self-adjusting state cannot be
/// meaningfully combined across multiple instances.
/// </remarks>
public sealed class SALSA : NativeSketch
{
    private readonly double _confidenceMetric;

    /// <summary>
    /// Creates a new SALSA sketch with the specified confidence metric.
    /// </summary>
    /// <param name="confidenceMetric">Confidence metric controlling the sketch behavior.
    /// Must be greater than 0. Higher values provide more confidence in estimates.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if confidenceMetric is not positive.</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public SALSA(double confidenceMetric)
    {
        if (confidenceMetric <= 0)
            throw new ArgumentOutOfRangeException(nameof(confidenceMetric), confidenceMetric, "Confidence metric must be greater than 0");

        _confidenceMetric = confidenceMetric;
        NativePtr = SketchOxideNative.salsa_new(confidenceMetric);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native SALSA");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private SALSA(double confidenceMetric, nuint ptr)
    {
        _confidenceMetric = confidenceMetric;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the confidence metric of this sketch.
    /// </summary>
    public double ConfidenceMetric
    {
        get
        {
            CheckAlive();
            return _confidenceMetric;
        }
    }

    /// <summary>
    /// Updates the sketch with a new element, incrementing its count.
    /// </summary>
    /// <param name="data">The bytes representing the item to count.</param>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                SketchOxideNative.salsa_update(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Updates the sketch with a string element, incrementing its count.
    /// </summary>
    /// <param name="value">The string item to count.</param>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Update(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Estimates the frequency count of an element.
    /// </summary>
    /// <param name="data">The bytes representing the item to query.</param>
    /// <returns>The estimated count for the item.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public ulong Estimate(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.salsa_estimate(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Estimates the frequency count of a string element.
    /// </summary>
    /// <param name="value">The string item to query.</param>
    /// <returns>The estimated count for the item.</returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public ulong Estimate(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Estimate(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Serializes the sketch to a byte array.
    /// </summary>
    /// <returns>Serialized sketch bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.salsa_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a SALSA sketch from a byte array.
    /// </summary>
    /// <param name="data">Serialized sketch bytes.</param>
    /// <param name="confidenceMetric">The confidence metric used when creating the original sketch.</param>
    /// <returns>A new SALSA instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static SALSA Deserialize(byte[] data, double confidenceMetric)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.salsa_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize SALSA: invalid data");

        return new SALSA(confidenceMetric, ptr);
    }

    /// <summary>
    /// Returns a string representation of the sketch.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "SALSA(disposed)";
        return $"SALSA(confidenceMetric={_confidenceMetric})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.salsa_free(NativePtr);
            NativePtr = 0;
        }
    }
}
